actions/cast_vote.js

```diff
@@ -4,39 +4,48 @@ const Shapes = require('../lib/shapes')
 const { fetchProp } = require('../lib/utils')
 const { ValidationError, NotFoundError } = require('../lib/errors')
 const CastVoteService = require('../services/cast_vote')
-const GetVoteStatsForPollService = require('../services/get_vote_stats_for_poll')
+const GetVoteStats = require('../services/get_vote_stats')
+const SavePollRating = require('../services/save_poll_rating')
 const Reply = require('../lib/reply')
 
 module.exports = function (opts = {}) {
-  this.add('sys:vote,vote:*', async function(msg, reply) {
+  this.add('sys:vote,vote:*', async function (msg, reply) {
     try {
-      // TODO:
-      // - [ ] Use seneca-joi
-      //
-      const validateMessage = Shapes.makeValidator(joi => joi.object({
-        vote: joi.string().valid(Vote.TYPE_UP(), Vote.TYPE_DOWN()).required(),
-        fields: joi.object({
-          poll_id: joi.string().max(64).required(),
-          voter_id: joi.string().max(64).required(),
-          voter_type: joi.valid('sys/user').required()
-        }).required()
-      }).unknown(), { stripUnknown: true })
-
       const vote_type = fetchProp(msg, 'vote')
       const safe_params = await validateMessage(msg)
       const poll_id = fetchProp(safe_params, ['fields', 'poll_id'])
       const voter_id = fetchProp(safe_params, ['fields', 'voter_id'])
       const voter_type = fetchProp(safe_params, ['fields', 'voter_type'])
+      const vote_kind = fetchProp(safe_params, ['fields', 'kind'])
+      const vote_code = fetchProp(safe_params, ['fields', 'code'])
+
+      // TODO: Implement transactions.
+      //
```

Comment:
For example, currently if the plugin crashes when trying to de-normalize a poll rating, the submission of the vote will not be rolled back (i.e. undone). This violates ACID and must be handled via implementation of a transaction.